### DISCUSSION TITLE
Implementing speech synthesis

### DISCUSSION THREAD

**Discussion by AstridOttmann** [neuefische/questions#497] [Web] [answered]
I want to implement speech synthesis in my app. I created a component (SpeechSynthesis), in which the functionality is handled.  I pass this to the component SingleEntry, which again is passed to the page ([id]. 
I get no error message, but the button is not working. 
I think, everything is handed over correctly, so I don't know where the error occurred. 
Here's the link to the corresponding branch [branch userstory-10](https://github.com/AstridOttmann/capstone-project/pull/21/files)

**Comment by doemser** [comment]
Your link leads me to a 404 page :(

**Reply by AstridOttmann** [reply]
Fixed it ;-)

**Reply by doemser** [reply]
If you write it like this:

```javascript
function handlePlaySpeech(voice, word) {
    const selectedVoice = availableVoices.find(
      (voice_) => voice_.name === voice
    );
    handleSpeech(word, selectedVoice);
  }

  return <SpeakerButton onClick={handlePlaySpeech} />;

```

the `voice` parameter will be filled with the event, which should be the cause for your error.

Right now your handlePlaySpeech function are no values passed, this is where you should start looking.

If you want it to say "Hallo" in german it should get  parameters  like the following passed:

```javascript

  return <SpeakerButton onClick={() => handlePlaySpeech("Google Deutsch", "Hallo")} />;
```

**Comment by AstridOttmann** [comment]
Thanks!
Tried to do it as you suggested, but then I get the error 'voice is not defined' 

` function handlePlaySpeech(voice, word) {
    const selectedVoice = availableVoices.find(
      (voice_) => voice_.name === voice
    );
    handleSpeech(word, selectedVoice);
  }

  return <SpeakerButton onClick={() => handlePlaySpeech(voice, word)} />;
}
`

[branch-userstory-10-changed](https://github.com/AstridOttmann/capstone-project/pull/21/files)

**Reply by AstridOttmann** [reply]
Also tried this: 
` function handlePlaySpeech(voice, word) {
    const selectedVoice = availableVoices.find(
      (voice_) => voice_.name === voice
    );
    handleSpeech(word, selectedVoice);
  }
  const selectedWord = translationList.filter((translation, { word }) => {
    return translation.word === word;
  });
  console.log(selectedWord.voice); // here I get undefined! Why?

  return (
    <SpeakerButton
      onClick={() => handlePlaySpeech(selectedWord.voice, selectedWord.word)}
    />
  );`

**Reply by AstridOttmann** [reply]
I found my mistake: I forgot to pass the variables 'word' and 'voice' to the SpeechSynthesis-component.

**Reply by doemser** [reply]
Nice you found it :) Maybe you can try to take a look at the custom hook approach as well, it is way better to let that useEffect only run once.

**Comment by doemser** [comment, marked as answer]
As far as I can see you have modified the prototype I originally sent you for speech synthesis in the direction that you attach the functionality to the button component and outsource it. Here is the original prototype converted to your use-case. So many changes should not be necessary for this.

Original Protoype: 
https://codesandbox.io/s/speech-synthesis-fl0nmj?file=/src/App.js

Your new use-case:
https://codesandbox.io/s/speech-synthesis-forked-pqvzdx?file=/src/App.js

**Reply by doemser** [reply]
It should be noted that you now execute the useEffect for all voices exactly as often as they are mapped, which is rather disadvantageous. It would be better to outsource the useEffect to collect the voices in a separate custom hook, for this there will be a workshop tomorrow, which will certainly help you.

**Reply by doemser** [reply]
Here with custom hook to it:

https://codesandbox.io/s/speech-synthesis-forked-forked-xe4tws?file=/src/App.js

**Reply by AstridOttmann** [reply]
Thanks a lot! :-)
I refactored my code, so that the useEffect is outsourced in a hook now. 
I left the functionality in the component, because I want to use it on several pages. 

I have one further question: 
in the prototype the handleSpeech-function is outside the component/App. I put it inside. What is the difference?

**Reply by doemser** [reply]
It is best to call the useVoices hook only once, not with every mapping, otherwise the advantage of the hook evaporates again. We only need to collect the voices once, so the useVoices hook is called intentionally outside the component you are mapping over and should be.

To your further question:

If a function doesn't need to be in the component, it shouldn't. Otherwise it will be rewritten every time the component is rendered. So if a function does not access useState logic or similar, you can usually take it out of the component and gain performance.
